storage/accumulator/src/lib.rs

```diff
@@ -185,30 +185,47 @@ where
         let root_level = Position::root_level_from_leaf_count(last_new_leaf_count);
         let mut to_freeze = Vec::with_capacity(Self::max_to_freeze(num_new_leaves, root_level));
 
-        // create one new node for each new leaf hash
-        let mut current_level = self.gen_leaf_level(new_leaves);
-        Self::record_to_freeze(
-            &mut to_freeze,
-            &current_level,
-            false, /* has_non_frozen */
-        );
-
-        // loop starting from leaf level, upwards till root_level - 1,
-        // making new nodes of parent level and recording frozen ones.
-        let mut has_non_frozen = false;
-        for _ in 0..root_level {
-            let (parent_level, placeholder_used) = self.gen_parent_level(&current_level)?;
-
-            // If a placeholder node is used to generate the right most node of a certain level,
-            // such level and all its parent levels have a non-frozen right most node.
-            has_non_frozen |= placeholder_used;
-            Self::record_to_freeze(&mut to_freeze, &parent_level, has_non_frozen);
+        let mut num_leaves = self.num_leaves;
+        let mut left_siblings: Vec<(_, _)> = Vec::new();
+        for leaf in new_leaves.iter() {
+            let leaf_pos = Position::from_leaf_index(num_leaves as u64);
+            let mut hash = *leaf;
+            to_freeze.push((leaf_pos, hash));
+            num_leaves += 1;
+            let mut pos = leaf_pos;
+            while pos.is_right_child() {
+                let sibling = pos.sibling();
+                match left_siblings.pop() {
+                    Some((x, left_hash)) => {
+                        assert_eq!(x, sibling);
+                        hash = Self::hash_internal_node(left_hash, hash);
+                    }
+                    None => hash = Self::hash_internal_node(self.reader.get(sibling)?, hash),
+                }
+                pos = pos.parent();
+                to_freeze.push((pos, hash));
+            }
+            left_siblings.push((pos, hash));
+        }
 
-            current_level = parent_level;
+        let (mut pos, mut hash) = left_siblings.pop().expect("Must have at least one node");
+        for _ in pos.level()..root_level as u32 {
```

Comment:
This could be re-written more concisely as a fold() - I don't see an easy way to do that with either of the loops above.

Reply:
They're fixpoints, but in absence of either a bound on call depth (or TCO in Rust) what you're doing is cleaner.